### PR TITLE
Reserve pool logic

### PR DESCRIPTION
**Pull request by almaraz97**
add deposit, withdraw, and getting of yield for the LP along with accounting of coverage amounts during the liquidity providing period.

### CODE REVIEW

File: contracts/src/modules/Coverage.sol [hunk 1]
```diff
@@ -59,28 +60,40 @@ contract Coverage is Ownable, ModuleBase {
...
+            liquidityReleaseDate = block.timestamp + liquidityLockupPeriod;
+        } else {
+            require(maturityDate <= liquidityReleaseDate);
+        }
 
         (address holder, uint256 amount, uint256 riskScore) = abi.decode(
             initData,
             (address, uint256, uint256)
-        );
+        );  // TODO do we need the risk score right now?
```
Comment (soolaymahn) [inline contracts/src/modules/Coverage.sol:78]: We can keep it in the contract and hard code 50 bps in the API for now

File: contracts/src/modules/Coverage.sol [hunk 2]
```diff
@@ -91,6 +104,58 @@ contract Coverage is Ownable, ModuleBase {
...
+        
+        availableFunds += coverage.coverageAmount;
+    }
+
+    function withdraw() public {
+        // HACK: if the total supply isn't claimed the module is bricked (can't restart LP/Coverage pool)
+        require(liquidityReleaseDate >= block.timestamp);  // Yielding period is over, allow claims
+        uint256 liquidityClaim = balanceOf(_msgSender());
+        uint256 claimPercentage = liquidityClaim / totalFunded;  // Percentage of yield they are entitled to
+        uint256 yieldClaim = yieldedFunds / claimPercentage; // TODO use safest math (round down)
```
Comment (soolaymahn) [inline contracts/src/modules/Coverage.sol:134]: should this be 
`uint256 yieldClaim = yieldedFunds * claimPercentage`?

File: contracts/src/modules/Coverage.sol [hunk 3]
```diff
@@ -59,28 +60,40 @@ contract Coverage is Ownable, ModuleBase {
...
             (address, uint256, uint256)
-        );
+        );  // TODO do we need the risk score right now?
 
         require(instant == (amount / 10000) * riskScore, "Risk fee not paid");
-        // TODO: maybe onramp should own nota? (currently the registrar assumes that the owner is the one being paid)
-        require(_owner == address(this), "Risk fee not paid to pool");
-        require(currency == usdc, "Incorrect currency");
+        
+        availableFunds -= amount;  // Throws on underflow (using pragma >0.8)
```
Comment (soolaymahn) [inline contracts/src/modules/Coverage.sol:82]: This assumes that the pool is full collateralized but with the coverage model the pool is under-collateralized. For Stably we're planning to offer 200k of coverage backed by 10k in the pool. Maybe a have a variable to track the `reserveRatio`?

File: contracts/src/modules/Coverage.sol [hunk 4]
```diff
@@ -6,81 +6,105 @@ import {DataTypes} from "../libraries/DataTypes.sol";
...
+        require(instant == (amount / 10_000) * riskScore, "Risk fee not paid");
+
+        uint256 currentCoverage = (totalReserves - availableReserves);
+        uint256 newCoverage = currentCoverage + amount;
+
+        uint256 newReserveBPS = (totalReserves / newCoverage) * 100;  // 1_000 / 800 => 800 * 100 => 8_000
+        if (newReserveBPS < reserveRatio) revert("Exceeds reserve ratio");
+
+        uint256 scaledAmount = (amount * reserveRatio) / 10_000;
+        availableReserves -= scaledAmount; // scale down by reserve ratio (amount=100 => amount=5 if rR=2_000)
```
Comment (soolaymahn) [inline contracts/src/modules/Coverage.sol:93]: hmmmm why not just do 
```
availableReserves -= amount * reserveRatio
if (availableReserves < 0) revert("Exceeds reserve ratio");
```

Comment (almaraz97) [reply contracts/src/modules/Coverage.sol]: The ratio is in BPS since decimals aren't well supported

**Review by soolaymahn — APPROVED** [review]
Fix the merge conflicts but otherwise looks good

**Pull request merged**